File: LeetCode/cpp/890_Find_and_Replace_Pattern.cpp
class Solution {
public:
	vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
		vector<string> answer;

		unordered_map<char, char> um1, um2;

		for (string word : words) {
			for (int i = 0; i < word.length(); i++) {
				um1[word[i]] = pattern[i];
				um2[pattern[i]] = word[i];
			}

			string w1, w2;

			for (int i = 0; i < word.length(); i++) {
				w1 += um1[word[i]];
				w2 += um2[pattern[i]];
			}

			if (w1 == pattern && w2 == word) {
				answer.push_back(word);
			}
		}

		return answer;
	}
};

File: LeetCode/java/890_Find_and_Replace_Pattern.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> answer = new ArrayList<>();

        Map<Character, Character> map1 = new HashMap<>();
        Map<Character, Character> map2 = new HashMap<>();

        for (String word : words) {
            for (int i = 0; i < word.length(); i++) {
                map1.put(word.charAt(i), pattern.charAt(i));
                map2.put(pattern.charAt(i), word.charAt(i));
            }

            StringBuilder w1 = new StringBuilder();
            StringBuilder w2 = new StringBuilder();

            for (int i = 0; i < word.length(); i++) {
                w1.append(map1.get(word.charAt(i)));
                w2.append(map2.get(pattern.charAt(i)));
            }

            if (w1.toString().equals(pattern) && w2.toString().equals(word)) {
                answer.add(word);
            }
        }

        return answer;
    }
}